### ISSUE TITLE
Backup feature?

### ISSUE THREAD

**Issue by buzucan** [WordOps/WordOps#15]
I didn't tested but I think that everyone needs backup. So, it will be nice to have some option to backup and upload in cloud. Yes, on wordpress are so many options but... it will be nice to have some included.

I saw a script here: https://github.com/openspace42/raptor maybe will be in WordOps too...

**Comment by tdtgit** [comment]
WO create a backup (source code + database) when we run `wo update ...`. Can create an sites array to do the samething at spectific time daily?

**Comment by VirtuBox** [comment]
Yes, but I also would like to implement MySQL backup with mariabackup (XtraDBackup fork), which provide the ability to perform hot and incremental MySQL backup without locking tables.
It require much less resources than mysqldump, and it provide the ability to restore the whole MySQL server in few minutes. 
For files backup, I'm still looking on solutions to perform incremental backup, because in my case, WordOps isn't installed on servers with a lot of storage available.

**Comment by juanpvh** [comment]
@virtubox all right ?!  I'm using wp-cli to do the backups.  would be a good tool to do both file and database backups in your opinion?

**Comment by tdtgit** [comment]
Brilliant, thanks for your comment. I will try mariabackup in my custom backup script tonight, to see how it's work.

For files backup, if you think we will run on low storage available server, so we should not make `--backup` option default but will ask every time users create a site. Then:
```
$ wo site create domain.com --backup
How many backup files you want to keep: 3
Choose your backup interval:
[1] Daily
[2] Weekly
[3] Monthly
3
Site created successful
...
```
Then we keep **one** original file and (**n - 1**) incremental backup files:
1. sitename.com.tar.gz
1. sitename.com.incre.27042019.tar.gz
1. sitename.com.incre.28042019.tar.gz

On the next day, we merge **sitename.com.tar.gz** with **sitename.com.incre.27042019.tar.gz**. I see tar [can do it](https://superuser.com/a/941552/759894), and maybe the others too. Then the structure will become:
1. sitename.com.tar.gz (merged)
1. sitename.com.incre.28042019.tar.gz
1. sitename.com.incre.29042019.tar.gz

**Comment by VirtuBox** [comment]
> @VirtuBox all right ?! I'm using wp-cli to do the backups. would be a good tool to do both file and database backups in your opinion?

If I remember properly, WP-CLI use mysqldump to perform backups.

**Comment by juanpvh** [comment]
> > @VirtuBox tudo bem ?! Estou usando o wp-cli para fazer os backups. Seria uma boa ferramenta para fazer backups de arquivos e bancos de dados na sua opinião?
> 
> Se bem me lembro, o WP-CLI usa o mysqldump para fazer backups.
> 
> @tdtgit muito boa ideia.

yes, I use more for the ease of creating the script and automating the process, using the --path = <path> to show where wp-config.php is, it informs database credentials, making the process easier.

> > @VirtuBox all right ?! I'm using wp-cli to do the backups. would be a good tool to do both file and database backups in your opinion?
> 
> If I remember properly, WP-CLI use mysqldump to perform backups.
> 
> @tdtgit very good idea.

yes, I use more for the ease of creating the script and automating the process, using the --path = <path> to show where wp-config.php is, it informs database credentials, making the process easier.

**Comment by aaronstuder** [comment]
Any updates on this? :)

**Comment by juanpvh** [comment]
> Alguma atualização sobre isso? :)

Add in v4.0

**Comment by michacassola** [comment]
Can we have other storage like Google Drive @VirtuBox ?

**Comment by mardoqueo** [comment]
> Can we have other storage like Google Drive @VirtuBox ?

+1000 on this.

**Comment by VirtuBox** [comment]
After running tests with several backup solutions, [restic](https://restic.net/) seems to be a good choice.
It support the most popular object storage providers, but it can also be integrated with rclone, giving users the choice for backup storage.

**Comment by juanpvh** [comment]
I use rclone + google drive. I enjoyed meeting this restic. It's hard to administer Restic + rclone + Gdrive so I researched and found this project https://github.com/astrada/google-drive-ocamlfuse.
So you can only have Restic + gdrive. I need to test in order to change.

**Comment by VirtuBox** [comment]
> I use rclone + google drive. I enjoyed meeting this restic. It's hard to administer Restic + rclone + Gdrive so I researched and found this project https://github.com/astrada/google-drive-ocamlfuse.
> So you can only have Restic + gdrive. I need to test in order to change.

Is it not very similar to rclone mount option ?

**Comment by juanpvh** [comment]
the configuration is direct with google drive client id and secret with restic. rlcone configures drive access with your information (can be changed). more in the end are similar. my question and have to use 3 tools :). rclone has no automated installation option yet making it impossible to develop scripts. Very manual work for many servers.

**Comment by michacassola** [comment]
I guess passing the necessary options together with the `--backup` command would be necessary. Maybe we can agree on the most important storage providers to optimize development and tools used?

**Comment by VirtuBox** [comment]
> I guess passing the necessary options together with the `--backup` command would be necessary. Maybe we can agree on the most important storage providers to optimize development and tools used?

I'm planning to make something similar to DNS API with acme.sh.
example with Minio : 
```bash
# setting credentials
export AWS_ACCESS_KEY_ID=<YOUR-MINIO-ACCESS-KEY-ID>
export AWS_SECRET_ACCESS_KEY= <YOUR-MINIO-SECRET-ACCESS-KEY>

# init backup storage
wo backup init localhost:9000/restic --minio

# backup site
wo site backup site.tld --minio

# set auto backup
wo site backup site.tld --minio --daily
```

But, to use the most popular storage providers (Google Drive, DropBox, OneDrive), it will require to use rclone somewhere, and rclone setup isn't very user friendly but there aren't a lot of alternatives

**Comment by maxwelljordanwhite** [comment]
@VirtuBox I am thrilled to see this happening. Staying tuned for V4. 👍

**Comment by VirtuBox** [comment]
> rclone has no automated installation option yet making it impossible to develop scripts. Very manual work for many servers.
I agree about that, but rclone config can be easily exported easily by creating an archive of the directory ~/.config/rclone

**Comment by pjv** [comment]
currently, i am backing up my WP sites using `mysqldump` and [tarsnap](https://tarsnap.com) driven by `cron` and custom bash scripts.

Last year I looked into various alternatives and narrowed down the candidates to two: [restic](https://restic.net/) and [duplicacy](https://duplicacy.com/home.html). 

My recollection is that in my testing, both seemed very good and each of them had one niggle (different in each case) that made them not 100% perfect to me and since my custom solution was already established and working, I've left the whole thing on the back burner and I can't remember anymore what the "niggles" were.

I just bring this up @VirtuBox in case you haven't seen duplicacy which is worth looking at. Since my last look at it, it seems that they have added a fairly extensive web-based GUI to manage backup jobs. That seems like it could be incorporated into the WO web admin dashboard and provide a user-friendly alternative to the somewhat difficult restic config.

**Comment by VirtuBox** [comment]
Thanks for this excellent suggestion @pjv, I wasn't aware there was a web-based GUI for duplicacy.

**Comment by VirtuBox** [comment]
Ah it was a little bit too beautiful..  duplicacy is now a paid software to perform backups : https://duplicacy.com/buy.html

**Comment by pjv** [comment]
> Ah it was a little bit too beautiful.. duplicacy is now a paid software to perform backups

oh... that's too bad.

**Comment by JPaulMora** [comment]
Can we just specify a backup folder? then i can just point [Duplicati](https://www.duplicati.com/) to it. 

I'd love to have binlogs stored too out of the box but i guess i can manage that from Duplicati

**Comment by aaronstuder** [comment]
> Can we just specify a backup folder? then i can just point [Duplicati](https://www.duplicati.com/) to it.
> 
> I'd love to have binlogs stored too out of the box but i guess i can manage that from Duplicati

Agreed :)

**Comment by michacassola** [comment]
@pjv Could we please have the backup script or better yet a tutorial with all the components?

@VirtuBox What is your timeline on the feature? Or maybe there could be an in between feature as @JPaulMora mentioned?

I want to use Mautic, but want to be able to back up a mysqldumb and the htdocs folder to not loose any data. Wordpress is not a problem as I can use one of the many backup plugins.

**Comment by pjv** [comment]
@michacassola [here you go](https://github.com/pjv/tarsnap-wordpress-backup).

**Comment by dingman** [comment]
I'd also like to see the ability to export to off-server locations, such as SFTP, Google Drive, S3, Dropbox, maybe there is a common library which can be used to speed up the delivery of this feature

**Comment by JPaulMora** [comment]
@dingman I just made a writeup of my backup process. I enabled MySQL Binlogs and installed Duplicati, which pushes backups to Dropbox. Hope this is useful for everyone else: [link to article](https://jpaulmora.github.io/2020/03/22/how-to-backup-your-linux-server-with-duplicati/) (any feedback is welcome)

**Comment by ibrahimhajjaj** [comment]
I would love to see a similar feature on WordOps.

**Comment by Endda** [comment]
> I would love to see a similar feature on WordOps.

same. I would like having a backup solution that can connect to my local FTP server for full site backups. that will allow me to remove the plugin I currently use so it can free up server resources and remove a possible attack vector

**Comment by Endda** [comment]
any updates about this @VirtuBox

**Comment by ddur** [comment]
@VirtuBox 

How about incremental (S)FTP backup as a solution for everyone?
Because FTP Server could be installed everywhere, free at home/office or remote.

Is it possible?
